### PR TITLE
[PHI] Optimize CUDA impl of `repeat_interleave_with_tensor_index` 

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
### PR Category

Performance Optimization

### PR Types

Performance 

### Description

Pcard-75624

1. 使用 cub 库代替 CPU for循环，加速index数组的构建过程，以此提升`repeat_interleave_with_tensor_index`算子在CUDA上的性能。
2. 优化 repeat_interleave 的python接口逻辑，在ndim=1时不需要进行flatten，减少少量API耗时

> 测试场景：`x = randn([16, 156000])`, `repeats = paddle.randint(0, 50)`, `axis=1`

``` log
# 优化前：[16, 3821944] paddle: 0.012788±0.001366s; torch: 0.001527±0.000077 s, pd_mem: 518.415 MB
# 优化后：[16, 3832242] paddle: 0.002294±0.000022s; torch: 0.001410±0.000012 s, pd_mem: 518.064 MB
```

**优化幅度约6x，耗时基本跟torch一个量级**

<details><summary>性能测试代码</summary>
<p>

``` py
import paddle
import torch
from ppsci.utils.misc import Timer
import time


for i in range(100):
    # paddle
    x_pd = paddle.randn([16, 156000])
    # x_pd.stop_gradient = False
    axis = 1
    repeats_pd = paddle.randint(0, 50, [x_pd.shape[axis]])

    with Timer("paddle", False) as t_pd:
        y_pd = paddle.repeat_interleave(x_pd, repeats_pd, axis)
        # exit(0)
    pd_mem = paddle.device.cuda.max_memory_allocated() / (1<<20)

    # torch
    x_pt = torch.from_dlpack(x_pd.detach())
    # x_pt.requires_grad = True
    repeats_pt = torch.from_dlpack(repeats_pd.detach())

    torch.cuda.synchronize()
    t = time.perf_counter()
    y_pt = torch.repeat_interleave(x_pt, repeats_pt, axis)
    torch.cuda.synchronize()
    t_pt = time.perf_counter() - t

    assert torch.allclose(torch.from_dlpack(y_pd), y_pt)
    print(f"{y_pd.shape} paddle: {t_pd.interval:.6f} s; torch: {t_pt:.6f} s, pd_mem: {pd_mem:.3f} MB")
```

</p>
</details>

### CODE REVIEW

File: paddle/phi/kernels/funcs/repeat_tensor2index_tensor.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline paddle/phi/kernels/funcs/repeat_tensor2index_tensor.cc:1]: 2022 -> 2025

**Review by zyfncg — APPROVED** [review]

File: paddle/phi/kernels/funcs/repeat_tensor2index_tensor.cc (hunk 1 above)
Comment (HydrogenSulfate) [reply paddle/phi/kernels/funcs/repeat_tensor2index_tensor.cc:1]: > 2022 -> 2025

好的，我下个PR改一下
